src/Common/src/TypeSystem/IL/Stubs/ForwardDelegateCreationThunk.cs

```diff
@@ -0,0 +1,117 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Interop;
+using Debug = System.Diagnostics.Debug;
+using Internal.TypeSystem.Ecma;
+
+namespace Internal.IL.Stubs
+{
+    /// <summary>
+    /// Thunk to create Delegates from native function pointer
+    /// </summary>
+    public partial class ForwardDelegateCreationThunk : ILStubMethod
+    {
+        private readonly TypeDesc _owningType;
+        private readonly MetadataType _delegateType;
+        private readonly InteropStateManager _interopStateManager;
+        private MethodSignature _signature;
+        
+
+        public ForwardDelegateCreationThunk(MetadataType delegateType, TypeDesc owningType, InteropStateManager interopStateManager)
+        {
+            _owningType = owningType;
+            _delegateType = delegateType;
+            _interopStateManager = interopStateManager;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return _owningType.Context;
+            }
+        }
+
+        public override bool IsNativeCallable
+        {
+            get
+            {
+                return false;
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get
+            {
+                return _owningType;
+            }
+        }
+
+        public MetadataType DelegateType
+        {
+            get
+            {
+                return _delegateType;
+            }
+        }
+
+        public override MethodSignature Signature
+        {
+            get
+            {
+                if (_signature == null)
+                {
+                    _signature = new MethodSignature(MethodSignatureFlags.Static, 0, 
+                        DelegateType,
+                        new TypeDesc[] {
+                            Context.GetWellKnownType(WellKnownType.IntPtr)
+                            });
+                }
+                return _signature;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return "ForwardDelegateCreationStub__" + DelegateType.Name;
+            }
+        }
+
+        public override MethodIL EmitIL()
+        {
+            var emitter = new ILEmitter();
+            ILCodeStream codeStream = emitter.NewCodeStream();
+            codeStream.EmitLdArg(0);
+
+            codeStream.Emit(ILOpcode.newobj, emitter.NewToken(
+                _interopStateManager.GetNativeWrapperDelegate(DelegateType)
```

Comment:
I think this is a p/invoke delegate wrapper, not a wrapper delegate. Otherwise you would read this as 'creating a delegate over delegate invoke'....